src/Session.php

```diff
@@ -0,0 +1,222 @@
+ *
+ * Each session has an in-memory attribute table, which can be used to store
+ * application-defined key/value pairs. A session's attribute table can be
+ * modified locally, as well as remotely by peers that have received a command
+ * request or notification from the session.
+ *
+ * The attribute table is versioned. Each revision of the attribute table is
+ * represented by the Revision interface.
+ *
+ * An optimistic-locking strategy is employed to protect the attribute table
+ * against concurrent writes. In order for a write to succeed, it must be made
+ * through a Revision value that represents the current (most recent) revision.
+ *
+ * Individual attributes in the table can be "frozen", preventing any further
+ * changes to that attribute.
+ */
+interface Session
+{
+    /**
+     * The session's unique identifier.
+     *
+     * @return SessionId The session's unique identifier.
+     */
+    public function id(): SessionId;
+
+    /**
+     * CurrentRevision returns the current revision of this session.
+     *
+     * @return Revision Current revision of this session.
+     *
+     * @throws InvalidRevisionException If the session has been closed, and revision is invalid.
+     */
+    public function currentRevision(): Revision;
+
+    /**
+     * Call sends a command request to the next available peer listening to the
+     * $namespace namespace and waits for a response.
+     *
+     * In the context of the call, the sessions owning peer is the "client" and
+     * the listening peer is the "server". The client and server may be the same
+     * peer.
+     *
+     * Both $command and $out are passed to the command handler on the server.
+     *
+     * Calls always use a deadline; if $context does not have a deadline, a
+     * timeout equal to Config::DEFAULT_TIMEOUT is used.
+     *
+     * If the call completes successfully, err is nil and in is the
+     * application-defined response payload sent by the server.
+     *
+     * IsFailure(err) returns true if the error is an application-defined
+     * failure. Failures are server-side errors that are part of the command's
+     * public API, as opposed to unexpected errors. If err is a failure, out
+     * contains the failure's application-defined payload; for this reason
+     * $out->close() must be called, even if err is non-nil.
+     *
+     * @param string $command Application-defined command name
+     * @param object $out     Application-defined request payload.
+     *
+     * @return mixed The repsonse of the invoked call.
+     *
+     * @throws ServerException   If the error occurred on the server.
+     * @throws ClientException   For application-defined failures.
+     * @throws NotFoundException If the Session has been closed and the command request can not be sent.
+     */
+    public function call(
+        Context $context,
+        string $namespace,
+        string $command,
+        Payload $out
+    ): Payload;
```

Comment:
Some general notes for all of these methods in `Session`:

- Payloads can be annotated as `mixed`.  There's no need for any `Payload` type
- In `Go`, I named payloads `$in` and `$out` because it's possible to name return values and Go likes short names. I think it's probably better to name the `$out` parameter as `$payload`, especially since it wont have a type hint.
- "application-defined" exceptions should be represented by `FailureException` (failures are a first-class member of the Rinq API) 
- "server-side" exceptions should be represented by a `CommandException` (this is more about consistency with Go)